### PR TITLE
Modernizing the code base and making it more compliant to common C++ style guides

### PR DESCRIPTION
**Pull request by JanVogelsang**
After working on the nest kernel for some time, I stumbled across multiple cases of either error-prone or nonmodern code that in some cases even led to hard-to-find bugs that could have been easily avoided if the code followed modern C++ style guides.  
Running cang-tidy on the whole codebase with the most important checks gave me ~8000 possible improvements. I took the time and went through all suggestions and applied the most important ones, especially those for which I knew they would definitely not break anything. I manually applied about 2000 of them, fixing the following problems:
- Clean up _.gitignore_, as there were duplicated entries
- Remove an unused closing html tag in the SLI-documentation
- Annotate constructors with a single argument with the _explicit_ keyword, to make sure that there is no implicit conversion happening between types that should not be converted without explicitly doing so
- Remove redundant _void_ from method signature if the method accepts no arguments. Putting _void_ into the argument list is an artifact from old C code, but is not written in modern C++ anymore
- Remove explicit variable initialization in cases where the value is never used, as the variable is defined again before ever accessing its value
- Replace all old C _NULL_ pointers values by the modern C++ _nullptr_, which is advised by Bjarne Stroustrup himself
- Instead of initializing null-pointers using _type* ptr = 0;_, explicitly set the pointer to a _nullptr_ (_type* ptr = nullptr;_)
- Add _override_ annotation when a function is supposed to be overridden to make sure errors are prevented when the signature of the derived class' method signature deviates from the parent's signature. There has been a bug in the archiving node related to this issue for example, which would not have been in the code for many months if we had the _override_ annotation on that function

It still would be a lot of work to clean up the whole codebase, but this PR is a good start in making NEST a sustainable software.

### CODE REVIEW

**Comment by JanVogelsang** [comment]
The MacOS test seems to fail for some reason, can someone help me find the cause of the issue?
This was the only error I was able to find in the logs, but I have no idea what that is supposed to tell me:
```
System call: unlink(2) /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T//ompi.Mac-1663857184846.501/pid.77232/1/vader_segment.Mac-1663857184846.501.242a0001.3
  Error:       No such file or directory (errno 2)
```

**Comment by heplesser** [comment]
@JanVogelsang I built under macOS now (with Clang 14 while Github runners have Clang 13, but should not make much of a difference), and did not run into any problems. I also noted two things in the FULL_MACOS build here:
- All tests pass, so the error message seems inconsequential. Let's see if it appears in the future again.
- In the FULL_MACOS build, there is a large number of compiler warnings, which all originate from a small number of lines and are not issued with master. I also get them when building on my machine. Here some links to the log:
    - https://github.com/nest/nest-simulator/actions/runs/3106295133/jobs/5033069063#step:8:560
    - https://github.com/nest/nest-simulator/actions/runs/3106295133/jobs/5033069063#step:8:571
    - https://github.com/nest/nest-simulator/actions/runs/3106295133/jobs/5033069063#step:8:581
    - https://github.com/nest/nest-simulator/actions/runs/3106295133/jobs/5033069063#step:8:591
    - https://github.com/nest/nest-simulator/actions/runs/3106295133/jobs/5033069063#step:8:1159

and there are probably a few more. They all seem to be related to overriding.

**Comment by JanVogelsang** [comment]
@heplesser I addressed all issues related to overrides in my latest commit, which should fix the compiler warnings. The warning we got originated from the following problem:  
*'...' overrides a member function but is not marked 'override'*: Adding the override specifier is not mandatory and simply produces a compilation error whenever a function in a derived class that is supposed to override a function in the base class doesn't actually override anything. This helps us developers to avoid problems where the function signatures just vary slightly (e.g. using 'SpikeEvent' in one place and 'Event' in another, which would - and actually did - result in a bug).

**Comment by heplesser** [comment]
@JanVogelsang Thanks a lot, nice that all now passes without warnings on all systems. I saw that the "unlink" error message is still there on macOS, but that might just as well come from the py-mpi-testing setup, so I would not worry about it.

**Comment by JanVogelsang** [comment]
I went through all my changes again, to make sure that my statement that nothing changes performance-wise is correct, and that is indeed the case.  
For any potential reviewers: 90% of the changes are just missing _override_ specifiers and using _nullptr_ instead of _NULL_ or _0_ now. If somebody would review those, there are only about 100 lines left with other changes that might need some more time to review.

**Comment by hakonsbm** [comment]
There are still quite a few formatting errors. Most of them because of leftover whitespace in empty parentheses of functions, for example here:
https://github.com/nest/nest-simulator/blob/96620fe9d6f80dcc0b94a17fbe6bd2d52ff06049/libnestutil/lockptr.h#L124

It's not picked up because of a bug (#2471) which will be fixed by #2478.

File: sli/slimath.cc
Comment (heplesser) [inline sli/slimath.cc:1124]:
```suggestion
  op1->get() = ( op1->get() or op2->get() );
```

File: sli/tarrayobj.cc
Comment (heplesser) [inline sli/tarrayobj.cc:88]: I am amazed that this compiles without `;`.  I would definitely have the semicolon after the assert to avoid confusion of the reader.

File: sli/token.h
Comment (heplesser) [inline sli/token.h:213]: Remove if not used.

File: sli/token.h
Comment (heplesser) [inline sli/token.h:349]:
```suggestion
    return ( p != nullptr ) and p->equals( &d );
```

File: sli/tokenarray.h
Comment (heplesser) [inline sli/tokenarray.h:429]: Delete if not needed.

File: sli/tokenarray.h
Comment (heplesser) [inline sli/tokenarray.h:443]: Delete if not needed.

File: sli/triedatum.cc
Comment (heplesser) [inline sli/triedatum.cc:39]:
```suggestion
  return fd != nullptr and tree == fd->tree;
```

File: sli/typechk.cc
Comment (heplesser) [inline sli/typechk.cc:71]:
```suggestion
  if ( next == nullptr and alt == nullptr ) // Leaf node
```

File: sli/typechk.cc
Comment (heplesser) [inline sli/typechk.cc:95]:
```suggestion
  if ( next == nullptr and alt == nullptr ) // Leaf node
```

**Review by heplesser — CHANGES_REQUESTED** [review]
@JanVogelsang An impressive effort, thank you! I have some comments, please see inline.

**Comment by heplesser** [comment]
@jougs Could you be the other reviewer? 95+% is quickly reviewed boiler plate, in the remaining 5% are a few things that would benefit from experienced eyes.

**Comment by heplesser** [comment]
@JanVogelsang Just one more comment: In order to keep this PR manageable, I'd suggest to not touch any other code lines in this PR beyond those already touched (except adding `#include <cmath>` if needed, see inline comments). For future similar PRs, I'd suggest to strictly separate changes if at all possible, e.g., one PR only adding `override`,  one only changing to `nullptr` etc to make reviewing easier.

File: libnestutil/beta_normalization_factor.h
Comment (JanVogelsang) [reply libnestutil/beta_normalization_factor.h:78]: There is no need to initialize the variable here, as the value is never used before it is set to another value in the following lines.

File: nestkernel/exceptions.h
Comment (JanVogelsang) [reply nestkernel/exceptions.h:655]: There is no need to initialize the std::string, as it will get default-constructed and initialized to an empty string (https://cplusplus.com/reference/string/string/string/).

File: nestkernel/music_manager.cpp
Comment (JanVogelsang) [reply nestkernel/music_manager.cpp:37]: I will fix all those in another PR, removing all unused includes that I can find.

File: nestkernel/proxynode.h
Comment (JanVogelsang) [reply nestkernel/proxynode.h:115]: A static function is (ever so slightly) more performant than a regular one, as a static function does not receive the implicit _this_ parameter, which is one less pointer to be sent around.

File: sli/dictutils.h
Comment (JanVogelsang) [reply sli/dictutils.h:93]: The variable x is not used anwhere before it gets initialized.

File: sli/dictutils.h
Comment (JanVogelsang) [reply sli/dictutils.h:157]: The variable x is not used anwhere before it gets initialized.

File: libnestutil/beta_normalization_factor.h
Comment (heplesser) [reply libnestutil/beta_normalization_factor.h:78]: That is true for the code as it is now, but it may change in the future. But initializing a factor with 0 seems dangerous, 1 might be a better value. But looking at the code of the entire function, I would suggest to remove the `normalization_factor` variable entirely and just return directly from the if-else below. Since that is the last statement of the function, I think that is sufficiently clean.

**Comment by JanVogelsang** [comment]
@heplesser Re-requesting a review seems to remove other reviewers from the list, could you add @jougs again?

File: nestkernel/proxynode.h
Comment (heplesser) [reply nestkernel/proxynode.h:115]: I would argue that it increases code complexity a little (programmers need to keep in mind that it is static), so unless there is a demonstrable gain in performance, I'd advise against `static` here.

But this makes me wonder if there is not another problem: `Node::get_thread()` is *not* `virtual`, but `proxynode`, which derives from `Node`, defines its own `get_thread()`. Can this be right?

File: sli/dictutils.h
Comment (heplesser) [reply sli/dictutils.h:93]: I would still argue to initialize, in case code changes later. One could consider to initialize with NaN to clearly mark this as a "dummy" value.

**Review by heplesser — CHANGES_REQUESTED** [review]
Thanks for the update, please see the inline comments I left.

File: nestkernel/proxynode.h
Comment (JanVogelsang) [reply nestkernel/proxynode.h:115]: Well, for programmers nothing really changes, as the function can still be called just like before. I can just make it `const` though, as well, the performance gain would be neglectable anyway.

I had a discussion with @med-ayssar regarding this issue. We came to the conclusion that `virtual` is indeed not needed in many cases where classes derive from `Node` (e.g. `set_spiketime`). However, in this case we need the virtual specifier, as the `get_thread` function will indeed be called on derived classes from the `Node::` interface and therefore the "overriden" implementation would not be used in these cases.
@jougs Can you confirm this?

File: sli/dictutils.h
Comment (JanVogelsang) [reply sli/dictutils.h:93]: Valid point. I think either 0 or NaN would be fine here. Let's go for NaN to make it explicit that the value is not used anywhere. Where is the NaN defined we are using? I have seen it somewhere, but I can't remember and I can't find any (well, only a single one) usages of `std::nan`, so I guess we use some macro for NaNs, right?

Comment (heplesser) [reply sli/dictutils.h:93]:
```c++
#include <limits>
std::numeric_limits< double >::quiet_NaN()
```

**Comment by heplesser** [comment]
> I don't get why the static checks fail. It gives me the following error message:
> 
> ```
> Files with erroneous copyright headers:
> ... sli/sli/lockptrdatum.h
> ```
> 
> I can't see how the copyright header of that file is different to the other's though.

It seems that you accidentally moved `sli/dictstack.h` to `sli/sli/lockptrdatum.h`, so now the file name in the copyright notice and filename do not agree.

**Comment by heplesser** [comment]
> > I don't get why the static checks fail. It gives me the following error message:
> > ```
> > Files with erroneous copyright headers:
> > ... sli/sli/lockptrdatum.h
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I can't see how the copyright header of that file is different to the other's though.
> 
> It seems that you accidentally moved `sli/dictstack.h` to `sli/sli/lockptrdatum.h`, so now the file name in the copyright notice and filename do not agree.

See you found it too :)

**Comment by JanVogelsang** [comment]
Got it! No idea how that happened...
Anyway, with that out of the way, a new challenger appeared: `sli/lockptrdatum.h`. The pipeline complains that the formatting of this file is incorrect. However, if I run `clang-format-13` (which I think is the same that the static_checks pipeline runs) on that file, the check still tells me the formatting is incorrect. If I revert any changes that have been made to this function (to a version that passed the checks before), the checks still don't pass.  
I might overlook something here, but at this point I am genuinely confused.

File: sli/lockptrdatum.h
Comment (heplesser) [inline sli/lockptrdatum.h:109]: @JanVogelsang The `;` is missing at the end of this line and that thows clang-format off. Given that this definition was commented out before: Is it needed at all? If not, why not delete it including the comment above?

Comment (JanVogelsang) [reply sli/lockptrdatum.h:109]: You are right. I thought I was the one who commented out that function, should have checked that before I commented it back in.  
But as it was commented out before as well already, I will now remove it completely as you suggested.

**Comment by JanVogelsang** [comment]
Yeah, will do that every time before pushing in the future. Usually do that, didn't do it in this case as I didn't imagine the changes could break anything.  

Good to know, will run that!

**Comment by med-ayssar** [comment]
> @hakonsbm How do I run `build_support/format_all_c_c++_files.sh`? It gives me the error `clang-format not found`, however running `clang-format` outside the script works just fine.

Maybe better to not run that script directly, as it is being called from the `ci_build.sh`, which does some setups before running the whole check. 

I recommend to check instead `static_code_analysis`, as is the main script for checking the formatting.
It starts here: [cppchecks](https://github.com/nest/nest-simulator/blob/e1c80d01f8318b5b2ce5d2f292cafb1f0e26047d/build_support/static_code_analysis.sh#L195)

**Comment by JanVogelsang** [comment]
@hakonsbm By the way, is it possible to squash commits that are already pushed somewhere? Would be great if I could squash all commits of this PR into a single one once everything is done.

**Comment by heplesser** [comment]
I would think that squashing could make sense in this case, but I would leave the final word to @jougs and @terhorstd.

**Comment by hakonsbm** [comment]
@JanVogelsang @med-ayssar `format_all_c_c++_files.sh` is not being called from the CI, as that would change the files, but is meant as a tool for developers (see [Check your code](https://nest-simulator.readthedocs.io/en/stable/contribute/coding_guidelines_check.html#check-code) in the documentation). The script defaults to the command `clang-format`, but you can give a different command by setting the `CLANG_FORMAT` variable:
```bash
CLANG_FORMAT=clang-format-13 build_support/format_all_c_c++_files.sh
```

**Comment by JanVogelsang** [comment]
@hakonsbm I am unsure why the script is not working out of the box, as it seems correct to me (and `clang-format` does exist and can be found from within bash), but defining `CLANG_FORMAT=clang-format-13` manually does the job as well, so that will work.

File: sli/triedatum.cc
Comment (jougs) [inline sli/triedatum.cc:39]: Not sure this is clearer now. Why not just have an `if/else` with two explicit `return` statements

**Review by jougs — COMMENTED** [review]
I'm generally positive about this one even though the problem with this type of PR is usually not what is in, but what was forgotten and thus cannot be reviewed ;-)

I'll approve when my comments have been addressed

File: models/iaf_cond_beta.h
Comment (heplesser) [reply models/iaf_cond_beta.h:441]: A general comment: Only the person who raised a question should click "Resolved", in the sense of "I am satisfied with the answer given or change made".

File: sli/triedatum.cc
Comment (JanVogelsang) [reply sli/triedatum.cc:39]: In my opinion, this line should stay as is. The first part of the statement simply ensures `fd` is defined, so that the following check does not try to dereference a `nullptr`.  
It could of course be expanded to 
```cpp
if (fd != nullptr)
{
    return tree == fd->tree;
}
return false;
```
which would need 5 more lines while making is barely more explicit? Or did you have something else in mind?

File: models/iaf_cond_beta.h
Comment (JanVogelsang) [reply models/iaf_cond_beta.h:441]: Alright, makes sense! I will unresolve everything I marked as resolved.

File: libnestutil/beta_normalization_factor.h
Comment (JanVogelsang) [reply libnestutil/beta_normalization_factor.h:78]: Valid points! Cleaned it up.

File: models/iaf_cond_beta.h
Comment (JanVogelsang) [reply models/iaf_cond_beta.h:441]: Must have skipped this file by accident, but anyway, added overrides here as well now!

File: models/iaf_psc_exp_multisynapse.cpp
Comment (JanVogelsang) [reply models/iaf_psc_exp_multisynapse.cpp:178]: By the way, should we move from `&&` to `and` in the entire codebase? I can create a PR if you want

File: nestkernel/nest_time.h
Comment (JanVogelsang) [reply nestkernel/nest_time.h]: True, made those implicit by accident, after removing some other `explicit` specifiers I first added, but later decided to remove again (as not needed). Fixed this.

File: nestkernel/proxynode.h
Comment (JanVogelsang) [reply nestkernel/proxynode.h:115]: True, that helps!

File: models/iaf_psc_exp_multisynapse.cpp
Comment (heplesser) [reply models/iaf_psc_exp_multisynapse.cpp]: In principle yes, but let us discuss this in the next developer VC. We currently have many open PRs and such massive changes can cause some trouble there.

File: models/stdp_synapse_facetshw_hom.h
Comment (heplesser) [reply models/stdp_synapse_facetshw_hom.h]: @jougs The change is in now, so I'll mark as resolved.

File: nestkernel/archiving_node.h
Comment (heplesser) [reply nestkernel/archiving_node.h]: If I see the actual changes in `archiving_node.h` correctly, the original line 106 stays in. I think that is sensible, even if you should not be used at present.

File: nestkernel/exceptions.h
Comment (heplesser) [reply nestkernel/exceptions.h:655]: @jougs What do you think about the explicit vs implicit initialization here? @JanVogelsang is right that there is no need, but with some members in the initializer list and others not, keeping track is more difficult. But then we have IDEs for keeping track ...

File: nestkernel/music_manager.cpp
Comment (heplesser) [reply nestkernel/music_manager.cpp:37]: OK.

File: nestkernel/archiving_node.h
Comment (JanVogelsang) [reply nestkernel/archiving_node.h]: Right, I decided to comment it back in, as the function might be used in the future at some point, as you just said.

File: nestkernel/nest_time.h
Comment (heplesser) [reply nestkernel/nest_time.h]: @JanVogelsang I am confused. In my opinion, `explicit` should be here, but looking at the file under "Files changed", it is still missing.

Comment (JanVogelsang) [reply nestkernel/nest_time.h]: Strange, same for me. When looking at the file in the corresponding branch on my forked repo, the explicit specifiers are correctly set: https://github.com/JanVogelsang/nest-simulator/blob/refactoring/nestkernel/nest_time.h

File: nestkernel/proxynode.h
Comment (heplesser) [reply nestkernel/proxynode.h:115]: Thanks for the details, @jougs! I have thought it through once more, and the situation is thus:
- Methods in `Node` that deal only with fields in the `Node` base class, such as `get_thread()`, are plain (not virtual) methods, because they will work correctly even when used through a base class pointer or reference. 
- Having these methods non-virtual makes them faster.
- The only exception here is `proxynode`, on which `get_thread()` must never be called.
- `proxynode` defines its own `get_thread()`, thus hiding the one in `Node`. This means that
    - if we call `get_thread()` through `proxynode*`, we trigger the assertion
    - if we call it trough `Node*` or `ArchivingNode*`, we call `Node::get_thread()`. Are we sure this is right?
- No matter what, I would argue against the `static`—no gain and just potential confusion.

File: sli/dictutils.h
Comment (heplesser) [inline sli/dictutils.h:158]: A bit surprised this compiles, given that NaN is not defined for longs, only for floating point types. Here maybe `...::min` is better.

**Comment by heplesser** [comment]
@JanVogelsang Thanks, I have marked most of the matters I had raised as resolved.

File: nestkernel/nest_time.h
Comment (heplesser) [reply nestkernel/nest_time.h]: Strange indeed, I think Gitlab is confused. I will mark this as resolved.

**Comment by heplesser** [comment]
Concerning `proxynode::get_thread()`: I just tried (not in Jan's branch) to make `Node::get_thread()` virtual and have `proxynode::get_thread()` override it. If I do that, quite a number of tests fail in the testsuite. A minimal reproducer is 
```
SLI ] << /local_num_threads 4 >> SetKernelStatus
SLI ] /n /parrot_neuron 8 Create def
SLI ] /sr /spike_recorder 8 Create def
SLI ] n sr /one_to_one Connect
```
I have not systematically tried to go below four threads and eight neurons. The `assert(false)` triggers when `get_thread()` is called from https://github.com/nest/nest-simulator/blob/e1c80d01f8318b5b2ce5d2f292cafb1f0e26047d/nestkernel/connection_manager.cpp#L1448

Without the `virtual`, `source_thread` is assigned whatever `Node::get_thread()` here returns. This is not really problematic, because it so happens that the following code https://github.com/nest/nest-simulator/blob/e1c80d01f8318b5b2ce5d2f292cafb1f0e26047d/nestkernel/connection_manager.cpp#L1449-L1471

will in the end just return `NO_CONNECTION` if the node is a proxy.

But this does not seem clean to me. @jougs, what do you think?

**Comment by heplesser** [comment]
Just an update to my previous post: I think we can solve the problem by changing the order of test in the code above in such a way that we first check for `is_proxy()` and if that is true, we never ask for the thread. I will try to produce a PR for that later.

File: sli/dictutils.h
Comment (JanVogelsang) [reply sli/dictutils.h:158]: Was also wondering, but didn't look further into it, as it compiled without any issue.  
Unsure whether `::min` or simply `0` is more explicit here, but we can just go for `::min` as it at least makes the reader think about it and might come to the conclusion that the `::min` is just a placeholder value.

**Comment by JanVogelsang** [comment]
I am genuinely confused why, but after running `clang-tidy` again now (even though I am using clang-tidy v13 now instead of v15 to make it consistent with our clang-format version) I get 936 more occurrences of
* Missing overrides
* Using nullptr instead of NULL/0
* Redundant void argument
Should I fix those as well either in this PR after everything else is resolved or in a separate PR?

**Comment by jougs** [comment]
Please create another PR, as to reduce confusion for reviewers.

File: nestkernel/proxynode.h
Comment (JanVogelsang) [reply nestkernel/proxynode.h:115]: * That is not guaranteed, but is valid in this case. It might be possible that an implementation in a derived class does something different with the same fields of the base class compared to the implementation in the base class.
* Exactly. Static > Non-virtual const > Non-virtual > Virtual. Having _any_ virtual function in class will have a general impact on performance by the way, as having at least one virtual function will add a vtable to instances of the class with `sizeof(pointer)` (usually 4 or 8 bytes), which might have an impact on caching performance.
* The `get_thread` of the `proxynode` could very well also be deleted, the reason for its existence is only to see that the function does not exist for a reason (right?).
* Correct. Just tested this and yes, I get these results as well.
* Alright, will change the `static` back to `const`.

File: sli/triedatum.cc
Comment (heplesser) [reply sli/triedatum.cc:39]: I fear @jougs and I have different views on readability. Maybe I have written too much Python by now, but my preferred way of expressing this would be
```c++
return fd and fd->tree == tree;
```
Bjarne S himself writes "I prefer `if (p)` over `if (p != nullptr)` (§6.2.2, C++ Language, 4ed), so Python-style implicit conversion too bool is approved. And once one becomes used to this idiom (we should do it throughout in NEST ...), it is much more compact and with fewer degrees of freedom than

```c++
if ( fd )
{
  return tree == fd->tree;
}
else
{ 
  return false;
}
```
In the latter, I prefer an explicit `else` clause instead of falling of the end.

Comment (jougs) [reply sli/triedatum.cc:39]: Whatever. I can read both and it's deep in the SLI codebase. Let's just leave it as it is and get this merged.

File: nestkernel/proxynode.h
Comment (heplesser) [reply nestkernel/proxynode.h:115]: Thanks!

File: sli/triedatum.cc
Comment (JanVogelsang) [reply sli/triedatum.cc:39]: True, I also prefer leaving out the `!= nullptr`, much cleaner in my opinion. If you want, I can address those as well in the PR-sequel for the entire codebase.

Comment (jougs) [reply sli/triedatum.cc:39]: Fine by me.

File: nestkernel/exceptions.h
Comment (heplesser) [reply nestkernel/exceptions.h:655]: I checked Stroustrup and in §17.4.1 he explicitly states that "[i]f a member constructor needs no argument, the member need not be mentioned in the member initializer list" and provides an example close to ours. Only built-in types definitely need initialization because they otherwise have undefined values. So I rest my case here.

**Review by heplesser — APPROVED** [review]
@JanVogelsang Thanks for all the work! I have resolved all questions I had raised, so now just points raised by @jougs remain.

File: models/aeif_cond_alpha.h
Comment (jougs) [reply models/aeif_cond_alpha.h:222]: @JanVogelsang: can you please file an issue in [the NESTML repository](https://github.com/nest/nestml/issues/new) to remind yourself of this? I will now resolve this conversation trusting that that will happen.

**Review by jougs — APPROVED** [review]
I'm approving and closing all conversations, trusting that a NESTML issue will be filed soon. Many thanks for this nice cleanup!

**Pull request merged**
